fs/src/io_uring/memory.rs

```diff
@@ -14,66 +15,87 @@ use {
 // and chunk it in slices of up to 1G each.
 const FIXED_BUFFER_LEN: usize = 1024 * 1024 * 1024;
 
-pub enum LargeBuffer {
-    Vec(Vec<u8>),
-    HugeTable(PageAlignedMemory),
-}
+#[derive(thiserror::Error, Debug)]
+pub enum AllocError {
+    #[error("Out of memory")]
+    OutOfMemory,
 
-impl Deref for LargeBuffer {
-    type Target = [u8];
+    #[error("Invalid layout parameters : {0}")]
+    LayoutCreation(#[from] LayoutError),
 
-    fn deref(&self) -> &Self::Target {
-        match self {
-            Self::Vec(buf) => buf.as_slice(),
-            Self::HugeTable(mem) => mem.deref(),
-        }
-    }
+    #[error("Can't allocate a zero size buffer")]
+    LayoutSize,
+
+    #[error("Libc mmap failed")]
+    MMapFailed,
 }
 
-impl DerefMut for LargeBuffer {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        match self {
-            Self::Vec(buf) => buf.as_mut_slice(),
-            Self::HugeTable(mem) => mem.deref_mut(),
+impl From<AllocError> for io::Error {
+    fn from(error: AllocError) -> io::Error {
+        match error {
+            AllocError::OutOfMemory => io::Error::new(io::ErrorKind::OutOfMemory, "Out of memory"),
+            AllocError::LayoutCreation(layout_err) => {
+                io::Error::new(io::ErrorKind::InvalidInput, layout_err)
+            }
+            AllocError::LayoutSize => io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Can't allocate a zero size buffer",
+            ),
+            _ => io::Error::other(error),
         }
     }
```

Comment:
Why (or where) do we need an io::Error instead of AllocError?

Reply:
Yeah, hard to say which one is better since using straight up io::Error would mean a lot more wordiness at the location of the error. I don't think it really matters either way though lol. Like using io::Error directly would mean writing something like `io::Error::new(io::ErrorKind::InvalidInput, "can't allocate a zero size buffer")` inline instead of returning an enum.